metricproviders/datadog/datadog.go

```diff
@@ -142,18 +142,24 @@ func (p *Provider) parseResponse(metric v1alpha1.Metric, response *http.Response
 		return "", v1alpha1.AnalysisPhaseError, fmt.Errorf("Could not parse JSON body: %v", err)
 	}
 
-	if len(res.Series) < 1 || len(res.Series[0].Pointlist) < 1 {
-		return "", v1alpha1.AnalysisPhaseError, fmt.Errorf("Datadog returned no value: %s", string(bodyBytes))
+	var value *float64
+	if len(res.Series) > 0 && len(res.Series[0].Pointlist) > 0 {
+		series := res.Series[0]
+		datapoint := series.Pointlist[len(series.Pointlist)-1]
+
+		// If datadog series.Pointlist is not empty, set the pointer value, else let it stay nil
+		if len(datapoint) == 2 {
+			value = new(float64)
+			*value = datapoint[1]
+		}
 	}
 
-	series := res.Series[0]
-	datapoint := series.Pointlist[len(series.Pointlist)-1]
-	if len(datapoint) < 1 {
-		return "", v1alpha1.AnalysisPhaseError, fmt.Errorf("Datadog returned no value: %s", string(bodyBytes))
+	status, err := evaluate.EvaluateResult(value, metric, p.logCtx)
+	if value == nil {
+		return "0", status, err
```

Comment:
this is a magic number, ie. why `0` picked arbitrarily?

if there's a consistent convention we can tap into.. like.. "Argo Rollout queries default to 0 for metric interpolation," then this will be OK.

Reply:
The value returned here is what we claim was returned by the provider in json form. It's actually not used for evaluation, only to help users debug why their metric failed. I think it would be better to either:
* return `null` 
* return `res.Series` array serialized to json

This is better than a value 0, which will mislead the user in that the provider actually returned a response.

Reply:
So, there are 2 parts that actually return `0`.

1. The spot you caught, here in `datadog.go` that doesn't actually get used for calculation
2. Inside `utils/evaluate/evaluate.go`.
<img width="715" alt="Screen Shot 2021-09-29 at 2 45 11 PM" src="https://user-images.githubusercontent.com/11354884/135353025-86620c82-1e74-45d0-85ff-0c6171dd60af.png">

So the question is, if a nil pointer is provided (if an empty `result` is passed into `EvaluateResult`) what should happen?

The painful part is the datatype has to match whatever is provided in the expression. For example

```yaml
successCondition: isNil(result) || result < 0.05
```

The datatype of `result` in `result < 0.05` has to be `float`.

Reply:
I can create another function like.. 

```yaml
successCondition: compareWithNil(result, <, 0.05, true)
```

That allows `result` and `0.05` to be different datatypes (`*float` and `float`). If `result` is nil, fallback to the 4th argument to the default return value.

Reply:
Or, instead, a function like..

```yaml
successCondition: expressionFallback("result < 0.05", true)
```

That checks if `result` is nil. If nil, return the second argument. If not nil, run the expr library again.. (calls the expr lib twice in this case). I think this is heavier from a code/computation side, but maybe nicer from a usability perpsective

Reply:
Or, maybe...

```yaml
successCondition: default(result, 0) < 0.05
```

If result is nil, then use the default value for the calculation. This is probably the best approach honestly...

Reply:
I'm resolving this conversation with the latest push which includes `default()` functionality in the unit tests and the docs.